src/rgw/rgw_rados.cc

```diff
@@ -2466,42 +2466,57 @@ void RGWObjVersionTracker::generate_new_write_ver(CephContext *cct)
 
 
 int RGWPutObjProcessor_Append::prepare(RGWRados *store, string *oid_rand) {
-  //get manifest
-  RGWRados::Object target(store, s->bucket_info, *static_cast<RGWObjectCtx *>(s->obj_ctx), rgw_obj(s->bucket, s->object));
-  RGWRados::Object::Read stat_op(&target);
-  stat_op.params.obj_size = &cur_size;
-  stat_op.params.attrs = &attrs;
-
-  r = stat_op.prepare();
-  if (r < 0 && r != -ENOENT) {
+  RGWObjState *astate;
+  int r = store->get_obj_state(static_cast<RGWObjectCtx *>(s->obj_ctx), bucket_info, rgw_obj(s->bucket, s->object), &astate);
+  if (r < 0) {
     return r;
-  } else if (r == -ENOENT) {
+  }
+  cur_size = astate->size;
+  *cur_accounted_size = astate->accounted_size;
+  cur_manifest = &astate->manifest;
+  if (!astate->exists) {
     if (position != 0) {
       ldout(store->ctx(), 5) << "ERROR: Append position should be zero" << dendl;
-      return ERR_POSITION_NOT_EQUAL_TO_LENGTH;
+      return -ERR_POSITION_NOT_EQUAL_TO_LENGTH;
     } else {
-      first = true;
-      cur_part = 1;
+      cur_part_num = 1;
       head_obj.init(bucket, obj_str);
     }
   } else {
     // check whether the object appendable
     map<string, bufferlist>::iterator iter = attrs.find(RGW_ATTR_APPEND_PART_NUM);
-    if (iter == attrs.end()) {
+    if (iter == astate->attrset.end()) {
       ldout(store->ctx(), 5) << "ERROR: The object is not appendable" << dendl;
-      return ERR_OBJECT_NOT_APPENDABLE;
+      return -ERR_OBJECT_NOT_APPENDABLE;
     }
-    if (position != size) {
+    if (position != *cur_accounted_size) {
       ldout(store->ctx(), 5) << "ERROR: Append position should be equal to the obj size" << dendl;
-      return ERR_POSITION_NOT_EQUAL_TO_LENGTH;
+      return -ERR_POSITION_NOT_EQUAL_TO_LENGTH;
     }
-    head_obj = stat_op.state.obj;
-    target.get_manifest(&cur_manifest);
-    decode(cur_part, iter->second);
+    head_obj = astate->obj;
+    decode(cur_part_num, iter->second);
     cur_part_num ++;
-  }
+    //get the current obj etag
+    iter = astate->attrset.find(RGW_ATTR_ETAG);
+    if (iter != astate->attrset.end()) {
+      string s = rgw_string_unquote(iter->second.c_str());
+      size_t pos = s.find("-");
```

Comment:
I take it the append type always has AWS multipart-style etag, I didn't notice the matching code